### PR TITLE
dbeaver/dbeaver-ee#1104 fix filter save and restore for document databases

### PR DESCRIPTION
**Pull request by E1izabeth**

### CODE REVIEW

**Comment by Matvey16** [comment]
Filter is applied incorrectly after DBeaver restart

Initial filter: `info.rank IN (12,22,24,28)`

filter after restart: `rank IN (12,22,24,28)`

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/DataFilterRegistry.java [hunk 1]
```diff
@@ -176,18 +309,81 @@ private class ConfigSaver extends AbstractJob {
...
         }
+        
+        private Map<DBSAttributeBase, String> collectAttrsInfo(
+            @NotNull XMLBuilder xml, @NotNull SavedDataFilter sdf
+        ) throws IOException {
+            int counter = 0;
+            Map<DBSAttributeBase, String> attrsInfo = new LinkedHashMap<>();
+            for (DBDAttributeConstraintBase attrC : sdf.constraints.values()) {
+                if (attrC instanceof DBDAttributeConstraint) {
+                    counter = flattenAttributes(attrsInfo, counter, ((DBDAttributeConstraint) attrC).getAttribute());
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/DataFilterRegistry.java]: You may break here if counter > 0

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/DataFilterRegistry.java [hunk 2]
```diff
@@ -78,44 +183,70 @@ static class SavedDataFilter {
...
                 }
             }
             this.anyConstraint = dataFilter.isAnyConstraint();
             this.order = dataFilter.getOrder();
             this.where = dataFilter.getWhere();
         }
 
-        void restoreDataFilter(DBRProgressMonitor monitor, DBSDataContainer dataContainer, DBDDataFilter dataFilter) throws DBException {
+        void restoreDataFilter(
+            @NotNull DBRProgressMonitor monitor, @NotNull DBSDataContainer dataContainer, @NotNull DBDDataFilter dataFilter
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/DataFilterRegistry.java]: Please put each parameter on a new line

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/DataFilterRegistry.java [hunk 3]
```diff
@@ -60,12 +62,115 @@ class DataFilterRegistry {
...
+            return new RestoredAttributesInfo(
+                dataSource == null ? Collections.emptyMap() : bindToDataSourceImpl(savedFilter, dataSource));
+        }
+        
+        @NotNull
+        private static Map<RestoredAttribute, RestoredAttribute> bindToDataSourceImpl(
+                @NotNull SavedDataFilter savedFilter, @NotNull DBPDataSource dataSource
+        ) {
+            Map<RestoredAttribute, RestoredAttribute> boundAttrs = new HashMap<>();
+            Stack<RestoredAttribute> stack = new Stack<>();
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/DataFilterRegistry.java]: Never use Stack class. It is obsolete.
Use ArrayList

**Comment by Matvey16** [comment]
Verified

**Pull request merged**